jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConnection.java

```diff
@@ -11,7 +11,7 @@
 // ========================================================================
 //
 
-package org.eclipse.jetty.server.internal;
```

Comment:
Can you explain why this move is necessary?

Is it so that `ServletAPIRequest` can do:

```java
        HttpConnection httpConnection = (HttpConnection)_servletContextRequest.getConnectionMetaData().getConnection();
        httpConnection.getParser().servletUpgrade();
        outputStream.flush(); // commit the 101 response
        httpConnection.getGenerator().servletUpgrade(); // tell the generator it can send data as-is
```

We already have interfaces: `Connection.UpgradeFrom`, `Connection.UpgradeTo` to abstract away from specific connection types. so can't you use them instead? Get those methods to flip the parser and generator states?

Reply:
>Is it so that ServletAPIRequest can do:

Yes, this is why it was necessary.

> We already have interfaces: Connection.UpgradeFrom, Connection.UpgradeTo to abstract away from specific connection types. so can't you use them instead? Get those methods to flip the parser and generator states?

These interfaces are already implemented on `HttpConnection`, I think for the HTTP/2 upgrade.

So I think the options are (unless you have a better idea)
1. make `HttpConnection` public
2. create a new `Upgrade` interface and implement it on the `HttpConnection`
3. add a new method to the `Connection` interface.